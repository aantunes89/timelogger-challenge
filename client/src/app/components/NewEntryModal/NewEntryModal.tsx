import React, { useEffect, useState } from "react";

import { useProjects } from "../../hooks/useProjects";
import { useScreenState } from "../../hooks/useScreenState";
import { storeEntry } from "../../services/storageService";
import { CustomModal } from "../CustomModal/CustomModal";
import { FormSubmitEvent, FormInputEvent } from "../../types/FormEvents";

interface NewEntryModalProps {
  isOpen: boolean;
  onRequestClose: () => void;
}

export function NewEntryModal({ isOpen, onRequestClose }: NewEntryModalProps) {
  const { projectId } = useProjects();
  const { setShouldUpdate, setShowSnackBar, setSnackBarMsg } = useScreenState();

  const [description, setDescription] = useState<string>("");
  const [timeSpent, setTimeSpent] = useState<number>(0);
  const [hourlyRate, setHourlyRate] = useState<number>(0);
  const [totalPrice, setTotalPrice] = useState<number>(0);

  useEffect(() => {
    timeSpent && hourlyRate ? setTotalPrice(timeSpent * hourlyRate) : setTotalPrice(0);
  }, [timeSpent, hourlyRate]);

  function handleInputNumber({ target: { value } }: FormInputEvent, cb: (val: number) => void) {
    const newVal = value ? Number.parseInt(value) : 0;
    return cb(newVal);
  }

  async function onSaveEntry(event: FormSubmitEvent) {
    event.preventDefault();

    storeEntry({
      projectId,
      description,
      timeSpent,
      hourlyRate,
      totalPrice,
    });

    resetState();
    setShouldUpdate(true);
    setShowSnackBar(true);
    setSnackBarMsg("Successfully Saved");
  }

  function resetState() {
    setDescription("");
    setTimeSpent(0);
    setHourlyRate(0);
  }

  return (
    <CustomModal
      isOpen={isOpen}
      onRequestClose={onRequestClose}
      title="New Entry"
      disabled={totalPrice ? false : true}
      onSubmit={(e) => onSaveEntry(e)}
      totalPrice={totalPrice}
    >
      <>
        <div className="form-field">
          <label>Task Description</label>
          <input
            type="text"
            value={description}
            onChange={(event) => setDescription(event.target.value)}
          />
        </div>

        <div className="form-field">
          <label>Time Spent</label>
          <input
            type="number"
            min={0}
            value={timeSpent}
            onChange={(e) => handleInputNumber(e, setTimeSpent)}
          />
        </div>

        <div className="form-field">
          <label>Hourly Rate</label>
          <input
            type="number"
            min={0}
            value={hourlyRate}
            onChange={(e) => handleInputNumber(e, setHourlyRate)}
          />
        </div>
      </>
    </CustomModal>
  );
}
