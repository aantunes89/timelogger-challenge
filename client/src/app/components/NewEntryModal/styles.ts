import styled from "styled-components";

export const Container = styled.form`
  h2 {
    color: var(--text-title);
    font-size: 1.5rem;
    margin-bottom: 2rem;
  }

  input {
    width: 100%;
    padding: 0 1.5rem;
    height: 3rem;
    margin-top: 1.5rem;

    border-radius: 0.25rem;
    border: 1px solid #d7d7d7;

    background-color: #e7e9ee;

    font-weight: 400;
    font-size: 1rem;

    &::placeholder {
      color: var(--text-body);
    }

    & + input {
      margin-top: 1rem;
    }
  }

  button[type="submit"],
  button.cancel {
    width: 100%;
    height: 4rem;

    padding: 0 1.5rem;
    margin-top: 1.5rem;
    border-radius: 0.25rem;
    border: 0;

    background-color: var(--green);
    color: #fff;
    font-size: 1rem;
    font-weight: 600;

    transition: filter 0.4s;

    &:hover {
      filter: brightness(0.9);
    }
  }

  button.cancel {
    background-color: #d7d7d7;
  }

  .action-buttons__wrapper {
    display: flex;
    align-items: center;

    justify-content: space-between;

    button:first-child {
      margin-right: 0.5rem;
    }
  }
`;
