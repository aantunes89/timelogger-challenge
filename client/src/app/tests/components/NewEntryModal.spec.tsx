import { fireEvent, render } from "@testing-library/react";
import Modal from "react-modal";
import { NewEntryModal } from "../../components/NewEntryModal/NewEntryModal";
import { renderHook, act } from "@testing-library/react-hooks";
import { Dispatch, useState } from "react";
import React from "react";

describe("", () => {
  Modal.setAppElement(document.createElement("div"));

  it("should render NewEntryModal", () => {
    const { getByText } = render(<NewEntryModal isOpen={true} onRequestClose={jest.fn()} />);

    expect(getByText("New Entry")).toBeInTheDocument();
    expect(getByText("Task Description")).toBeInTheDocument();
    expect(getByText("Time Spent")).toBeInTheDocument();
    expect(getByText("Hourly Rate")).toBeInTheDocument();
  });

  it("should change task description input value", () => {
    const setTaskDescription = jest.fn();

    const { getByText } = render(<NewEntryModal isOpen={true} onRequestClose={jest.fn} />);

    const taskDescriptionInput = getByText("Task Description");
    taskDescriptionInput.addEventListener("change", setTaskDescription);

    fireEvent.change(taskDescriptionInput);

    expect(setTaskDescription).toHaveBeenCalled();
  });

  it("should change time spent input value", () => {
    const setTimeSpent = jest.fn();

    const { getByText } = render(<NewEntryModal isOpen={true} onRequestClose={jest.fn} />);

    const timeSpentInput = getByText("Time Spent");
    timeSpentInput.addEventListener("change", setTimeSpent);

    fireEvent.change(timeSpentInput);

    expect(setTimeSpent).toHaveBeenCalled();
  });

  it("should change hourly rate input value", () => {
    const setHourlyRate = jest.fn();

    const { getByText } = render(<NewEntryModal isOpen={true} onRequestClose={jest.fn} />);

    const hourlyRate = getByText("Hourly Rate");
    hourlyRate.addEventListener("change", setHourlyRate);

    fireEvent.change(hourlyRate);

    expect(setHourlyRate).toHaveBeenCalled();
  });

  it("should change hourly rate input value", () => {
    const setState = jest.fn();

    const useStateMock: any = (state: any) => [state, setState];
    jest.spyOn(React, "useState").mockImplementation(useStateMock);

    const { getByText } = render(<NewEntryModal isOpen={true} onRequestClose={jest.fn} />);
    const hourlyRate = getByText("Hourly Rate");
    hourlyRate.addEventListener("change", setState);

    fireEvent.change(hourlyRate);

    expect(setState).toHaveBeenCalled();
  });
});
