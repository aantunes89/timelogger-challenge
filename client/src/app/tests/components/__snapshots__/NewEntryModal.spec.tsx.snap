// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` should render NewProjectModal 1`] = `<div />`;
